app/models/position_opening.rb

```diff
@@ -1,134 +1,238 @@
-                },
-                :order => 'asc'
+        if tags
+          search_definition[:query][:bool][:must] << { terms: { tags: tags } }
+        end
+
+        if query.position_offering_type_code.present?
+          search_definition[:query][:bool][:must] << { match: { position_offering_type_code: { query: query.position_offering_type_code } } }
+        end
+
+        if query.position_schedule_type_code.present?
+          search_definition[:query][:bool][:must] << { match: { position_schedule_type_code: { query: query.position_schedule_type_code } } }
+        end
+
+        if query.keywords.present?
+          search_definition[:query][:bool][:should] << { match: { position_title: { query: query.keywords, analyzer: 'custom_analyzer' } } }
+        end
+
+        if query.keywords.present? && query.location.nil?
+          search_definition[:query][:bool][:should] << {
+            nested: {
+              path: 'locations',
+              query: {
+                match: { 'locations.city': { query: query.keywords, operator: 'AND' }}
               }
+            }
+          }
+        end
+
+        if query.rate_interval_code.present?
+          search_definition[:query][:bool][:must] << { match: { rate_interval_code: query.rate_interval_code }}
+        end
+
+        if query.organization_ids.present?
+          definition = {
+            bool: {
+              should: []
+            }
+          }
+
+          if query.organization_terms.present?
+            definition[:bool][:should] << { terms: { organization_id: query.organization_terms } }
+          end
+          if query.organization_prefixes.present?
+            query.organization_prefixes.each do |prefix|
+              definition[:bool][:should] << { prefix: { organization_id: prefix } }
             end
           end
+
+          search_definition[:query][:bool][:must] << definition
+        end
+
+        if query.location.present?
+          locations = {
+            nested: {
+              path: 'locations',
+              query: {
+                bool: {
+                  must: []
+                }
+              }
+            }
+          }
+
+          if query.has_state?
+            locations[:nested][:query][:bool][:must] << { term: { 'locations.state': { term: query.location.state } }}
+          end
+
+          if query.has_city?
+            locations[:nested][:query][:bool][:must] << { match: { 'locations.city': { query: query.location.city, operator: 'AND' }}}
+          end
+
+          search_definition[:query][:bool][:must] << locations
+        end
+      end
+
+      if query.keywords.blank?
+        if lat.blank? || lon.blank?
+          search_definition[:sort] = [{ "#{sort_by}": 'desc' }]
+        else
+          search_definition[:sort] = [{
+            _geo_distance: {
+              'locations.geo': { lat: lat.to_f, lon: lon.to_f },
+              order: 'asc',
+              nested_path: 'locations'
+            }
```

Comment:
nested_path is required when sorting nested fields: https://www.elastic.co/guide/en/elasticsearch/reference/5.2/search-request-sort.html#nested-sorting